src/librustc_interface/queries.rs

```diff
@@ -277,11 +279,57 @@ impl<'tcx> Queries<'tcx> {
                 // Don't do code generation if there were any errors
                 self.session().compile_status()?;
 
+                // Hook for compile-fail tests.
+                Self::check_for_rustc_errors_attr(tcx);
+
                 Ok(passes::start_codegen(&***self.codegen_backend(), tcx, &*outputs.peek()))
             })
         })
     }
 
+    /// Check for the `#[rustc_error]` annotation, which forces an error in codegen. This is used
+    /// to write compile-fail tests that actually test that compilation succeeds without reporting
+    /// an error.
+    fn check_for_rustc_errors_attr(tcx: TyCtxt<'_>) {
+        let def_id = match tcx.entry_fn(LOCAL_CRATE) {
+            Some((def_id, _)) => def_id,
+            _ => return,
+        };
+
+        const DELAY_SPAN_BUG_TRIGGERED: &str =
+            "delayed span bug triggered by #[rustc_error(delay_span_bug_from_inside_query)]";
```

Comment:
This is still not being triggered from inside a query, I don't understand.

Reply:
https://github.com/rust-lang/rust/pull/69965#discussion_r392678158

Reply:
> The name suggests the delay_span_bug would be triggered from inside a query, that's why there's a query that gets provided, but you removed all that.

Sorry, I'm a bit lost... Here's my understanding of this function:
- We want to check for `rustc_error` attributes specifically after all other analyses have finished 
- If a previous analysis failed, then we would never get here.
- If everything before this succeeds, then this error gets triggered unless the error is further delayed because it is a rustc_error(delayed)

Reply:
The message clearly says `#[rustc_error(delay_span_bug_from_inside_query)]`.
Isn't that the attribute? This is, if memory serves, for testing `delay_span_bug`, *specifically*. And through a query so that it hits a certain worst-case, wrt to incremental IIRC.

If it's not inside a query, it doesn't get the same incremental treatment, so the test will just pass forever without testing for the original problem at all.

Reply:
Ah, I see... that's not at all obvious from the code, and there are no comments... I will add back the query and a comment.

Reply:
> Ah, I see... that's not at all obvious from the code,

Yeah it could use some comments. But that's why it had `from_inside_query` in the name, to make it somewhat self-descriptive.